internal/impl/kafka/enterprise/redpanda_migrator_offsets_input.go

```diff
@@ -0,0 +1,207 @@
+		[]*service.ConfigField{
+			service.NewAutoRetryNacksToggleField(),
+		},
+	)
+}
+
+func init() {
+	err := service.RegisterBatchInput("redpanda_migrator_offsets", redpandaMigratorOffsetsInputConfig(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.BatchInput, error) {
+			tmpOpts, err := kafka.FranzConnectionOptsFromConfig(conf, mgr.Logger())
+			if err != nil {
+				return nil, err
+			}
+			clientOpts := append([]kgo.Opt{}, tmpOpts...)
+
+			d := kafka.FranzConsumerDetails{}
+
+			var topics []string
+			if topicList, err := conf.FieldStringList(rmoiFieldTopics); err != nil {
+				return nil, err
+			} else {
+				topics, _, err = kafka.ParseTopics(topicList, -1, false)
+				if err != nil {
+					return nil, err
+				}
+				if len(topics) == 0 {
+					return nil, errors.New("at least one topic must be specified")
+				}
+			}
+
+			var topicPatterns []*regexp.Regexp
+			if regexpTopics, err := conf.FieldBool(rmoiFieldRegexpTopics); err != nil {
+				return nil, err
+			} else if regexpTopics {
+				topicPatterns = make([]*regexp.Regexp, 0, len(topics))
+				for _, topic := range topics {
+					tp, err := regexp.Compile(topic)
+					if err != nil {
+						return nil, fmt.Errorf("failed to compile topic regex %q: %s", topic, err)
+					}
+					topicPatterns = append(topicPatterns, tp)
+				}
+			}
+
+			if d.RackID, err = conf.FieldString(rmoiFieldRackID); err != nil {
+				return nil, err
+			}
+
+			if d.FetchMaxBytes, err = kafka.BytesFromStrFieldAsInt32(rmoiFieldFetchMaxBytes, conf); err != nil {
+				return nil, err
+			}
+			if d.FetchMinBytes, err = kafka.BytesFromStrFieldAsInt32(rmoiFieldFetchMinBytes, conf); err != nil {
+				return nil, err
+			}
+			if d.FetchMaxPartitionBytes, err = kafka.BytesFromStrFieldAsInt32(rmoiFieldFetchMaxPartitionBytes, conf); err != nil {
+				return nil, err
+			}
+
+			// Consume messages from the `__consumer_offsets` topic
+			d.Topics = []string{`__consumer_offsets`}
+			clientOpts = append(clientOpts, d.FranzOpts()...)
+
+			rdr, err := kafka.NewFranzReaderOrderedFromConfig(conf, mgr, func() ([]kgo.Opt, error) {
+				return clientOpts, nil
+			}, func(record *kgo.Record) (*service.Message, error) {
+				msg := kafka.FranzRecordToMessageV1(record)
+
+				// Check the version to ensure that we process only offset commit keys
+				key := kmsg.NewOffsetCommitKey()
+				if err := key.ReadFrom(record.Key); err != nil || (key.Version != 0 && key.Version != 1) {
+					return nil, fmt.Errorf("failed to decode record key: %s", err)
+				}
+
+				isExpectedTopic := false
+				if len(topicPatterns) > 0 {
+					isExpectedTopic = slices.ContainsFunc(topicPatterns, func(tp *regexp.Regexp) bool {
+						return tp.MatchString(key.Topic)
+					})
+				} else {
+					isExpectedTopic = slices.ContainsFunc(topics, func(t string) bool {
+						return t == key.Topic
+					})
+				}
+				if !isExpectedTopic {
+					return nil, fmt.Errorf("skipping updates for the %q topic: %s", key.Topic, err)
```

Comment:
should we just `return nil, nil` for skipping records? I guess that's why in the previous commit you have the error case for this being to skip the record instead of raising an error.

Reply:
This way, I can `log.Debugf` higher up the reason for skipping messages. In general, I don't think the input should raise errors if it can't decode a message. Either drop it or send it as-is. We can, however, consider setting an error on the message itself, but I don't think any input does that currently and it might be a bit weird to do it. Curious if @Jeffail has any thoughts on this.